mcp/src/main/java/io/modelcontextprotocol/client/transport/StreamableHttpClientTransport.java

```diff
@@ -0,0 +1,396 @@
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import io.modelcontextprotocol.spec.McpClientTransport;
+import io.modelcontextprotocol.spec.McpSchema;
+import io.modelcontextprotocol.util.Assert;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.util.retry.Retry;
+
+/**
+ * A transport implementation for the Model Context Protocol (MCP) using JSON streaming.
+ *
+ * @author Aliaksei Darafeyeu
+ */
+public class StreamableHttpClientTransport implements McpClientTransport {
+
+	private static final Logger LOGGER = LoggerFactory.getLogger(StreamableHttpClientTransport.class);
+
+	private final HttpClientSseClientTransport sseClientTransport;
+
+	private final HttpClient httpClient;
+
+	private final HttpRequest.Builder requestBuilder;
+
+	private final ObjectMapper objectMapper;
+
+	private final URI uri;
+
+	private final AtomicReference<TransportState> state = new AtomicReference<>(TransportState.DISCONNECTED);
+
+	private final AtomicReference<String> lastEventId = new AtomicReference<>();
+
+	private final AtomicBoolean fallbackToSse = new AtomicBoolean(false);
+
+	StreamableHttpClientTransport(final HttpClient httpClient, final HttpRequest.Builder requestBuilder,
+			final ObjectMapper objectMapper, final String baseUri, final String endpoint,
+			final HttpClientSseClientTransport sseClientTransport) {
+		this.httpClient = httpClient;
+		this.requestBuilder = requestBuilder;
+		this.objectMapper = objectMapper;
+		this.uri = URI.create(baseUri + endpoint);
+		this.sseClientTransport = sseClientTransport;
+	}
+
+	/**
+	 * Creates a new StreamableHttpClientTransport instance with the specified URI.
+	 * @param uri the URI to connect to
+	 * @return a new Builder instance
+	 */
+	public static Builder builder(final String uri) {
+		return new Builder().withBaseUri(uri);
+	}
+
+	/**
+	 * The state of the Transport connection.
+	 */
+	public enum TransportState {
+
+		DISCONNECTED, CONNECTING, CONNECTED, CLOSED
+
+	}
+
+	/**
+	 * A builder for creating instances of WebSocketClientTransport.
+	 */
+	public static class Builder {
+
+		private final HttpClient.Builder clientBuilder = HttpClient.newBuilder()
+			.version(HttpClient.Version.HTTP_1_1)
+			.connectTimeout(Duration.ofSeconds(10));
+
+		private final HttpRequest.Builder requestBuilder = HttpRequest.newBuilder()
+			.header("Accept", "application/json, text/event-stream");
```

Comment:
I think we should put "Content-type: application/json" request header here. Some servers might not be able to process the request without this header.